Add a scheduler for background jobs

cevian opened the pull request:
TimescaleDB will want to run multiple background jobs. This PR
adds a simple scheduler so that jobs inserted into a jobs table
could be run on a schedule. This first implementation has two limitations:

1) The list of jobs to be run is read from the database when the scheduler
is first started. We do not update this list if the jobs table changes.

2) There is no prioritization for when to run jobs.

There design of the scheduler is as follows:
The scheduler itself is a background job that continuously runs and waits
for a time when  jobs need to be scheduled. It then launches jobs as new
background workers that it controls through the background worker handle.

Aggregate statistics about a job are kept in the job_stat catalog table.
These statistics include the start and finish times of the last run of the job
as well as whether or not the job succeeded. The next_start is used to
figure out when next to run a job after a scheduler is restarted.

The statistics table also tracks consecutive failures and crashes for the job
which is used for calculating the exponential backoff after a crash or failure
(which is used to set the next_start after the crash/failure). Note also that
there is a minimum time after the db scheduler starts up and a crashed job
is restarted. This is to allow the operator enough time to disable the job
if needed.

Note that the number of crashes is an overestimate of the actual number of crashes
for a job. This is so that we are conservative and never miss a crash and fail to
use the appropriate backoff logic.  Note that there is some complexity
in ensuring that all crashes are counted since a crash in Postgres causes /all/
processes to SIGQUIT: we must commit changes to the stats
table /before/ a job starts so that we can then deduce after a job has crashed
and the scheduler comes back up that a job was started, and not finished before
the crash (meaning that it could have been the crashing process).

amytai commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+
+static
+TimestampTz
+calculate_next_start_success(TimestampTz last_finish, BgwJob *job)
```
This function and `calculate_next_start_failure` (below) don't have to be 2 different functions. To make the logic easier, you can just have a `calculate_next_start` function that takes the FormData_bgw_job_stat and calculates the next start time by inspecting the values.

cevian replied:
I actually disagree. I think this way of splitting things up is more modular and makes it easier to test. (I do plan to add unit tests inside this module)

amytai commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+		fd->total_success++;
+		fd->consecutive_failures = 0;
+		/* Mark the next start at the end if the job itself hasn't */
+		if (!bgw_job_stat_next_start_was_set(fd))
```
If you put `calculate_next_start_success` and `calculate_next_start_crash` in the same function, then you only have to do this `if` statement once outside of the `if (res->success)` statement.

cevian replied:
as above

amytai commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+		 * Mark the next start at the end if the job itself hasn't (this may
+		 * have happened before failure)
+		 */
+		if (!bgw_job_stat_next_start_was_set(fd))
```
Coalesce with above

cevian replied:
as above

amytai commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,395 @@
+}
+
+void
+ts_bgw_scheduler_process(int32 ttl_ms)
```
Why does the scheduler take a TTL? What's a potential use-case for this? Shouldn't the scheduler just run forever until some signal?

cevian replied:
This is purely for making it easier to test. I'll add documentation.

amytai commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+	if (jobstat == NULL)
+		/* Never previously run */
+		return true;
+	return jobstat->fd.consecutive_crashes < MAX_CRASHES;
```
I thought the verdict was to remove a MAX_CRASHES check and just keep increasing the backoff?

cevian replied:
Good point. Fixed

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+
+	/* undo marking created by start marks */
+	fd->last_run_success = res->success;
+	fd->total_crashes--;
```
Hmm maybe this is a naming issue? The jobs table actually is the schedule of a job_type. So, it says how often it recurs, how often to run, etc. It is for jobs that always recur. I don't see much use for non-recuring jobs at this point. Maybe we should call that `job_schedule`? The job_stat table is the statistics for a particular job_schedule not sure it make sense for statics to be about a single instantiation (run) of a job. Or am I missing your point here?

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,34 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
 ON _timescaledb_catalog.chunk_index(hypertable_id, hypertable_index_name);
 SELECT pg_catalog.pg_extension_config_dump('_timescaledb_catalog.chunk_index', '');
 
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.bgw_job (
```
I was just going to use sql to upsert a row in. This can be done in the sql scripts and the upgrade script. We should do this after everything is merged.

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,248 @@
+bool
+ts_bgw_job_execute(BgwJob *job)
+{
+	if (namestrcmp(&job->fd.job_type, "check_for_update_if_enabled") == 0)
```
Unfortunately we currently don't have a way to have a common macro between sql and C. I think we can resolve this after the merge. For this string in C I will create a #define.

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,248 @@
+	BgwJob	  **job_pp = data;
+
+	*job_pp = bgw_job_from_tuple(ti->tuple, ti->mctx);
+	return false;
```
False means don't continue scan I believe.

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,344 @@
+	BgwJobStat **job_stat_pp = data;
+
+	*job_stat_pp = bgw_job_stat_from_tuple(ti->tuple, ti->mctx);
+	return false;
```
False means don't continue scan I believe

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,34 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+    job_timeout         INTERVAL    NOT NULL,
+    max_retries         INT         NOT NULL,
+    retry_period        INTERVAL    NOT NULL,
+    CONSTRAINT  valid_job_type CHECK (job_type IN ('update_check'))
```
Let's settle on a name soon. Discuss tomorrow?

mfreed commented:
Your PR comment also provides a great summary of the design of the job schedule.  Have you considered adding them as a high-level comment in the *code*, to be modified as some of the aforementioned limitations are removed?

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,34 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+    retry_period        INTERVAL    NOT NULL,
+    CONSTRAINT  valid_job_type CHECK (job_type IN ('update_check'))
+);
+SELECT pg_catalog.pg_extension_config_dump('_timescaledb_catalog.bgw_job', '');
```
Hmm yeah I agree the job_stat should not be dumped. Dumping the job table though.

cevian replied:
I think its weird if you restore a database and now all of your jobs are gone. In fact if I restored a db with a `drop chunks` job and the restored db did not drop the data the original did, I would find that surprising behavior. My opinion is: the restored DB should act, as much as possible, like the original database.

cevian commented on src/bgw/job_stat.h:
```diff
@@ -0,0 +1,24 @@
+bool		ts_bgw_job_stat_should_execute(BgwJobStat *jobstat, BgwJob *job);
+
+TimestampTz
+			ts_bgw_job_stat_next_start(BgwJobStat *jobstat, BgwJob *job);
```
fixed

cevian commented:
@mfreed Good point. Added a README in src/bgw.

@amytai I believe I addressed all you points, thanks for all your work reviewing. Ready for another review.

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+}
+
+static bool
+scheduled_bgw_job_is_running(ScheduledBgwJob *sjob)
```
I think this function name is misleading, ISTM this checks if the job was once started, regardless of whether it's running now.

cevian replied:
No it checks whether it's running now because scheduled_bgw_job_mark_stopped will set the handle to NULL.

JLockerman replied:
So compare this function to something like waitpid, that function will return information about the ground truth of the process state regardless of when and how it's called. This function returns information based on the last call to scheduled_bgw_job_mark_stopped, which, depending on where we are relative to scheduled_bgw_job_list_check_stopped_or_timed_out_jobs may or may not be up-to-date. A more accurate name would be something like scheduled_bgw_job_is_scheduled.

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+			/* all good */
+			break;
+		case BGWH_STOPPED:
+			sjob->handle = NULL;
```
note, we also null handle in `scheduled_bgw_job_mark_stopped`, if we need to do so here as well, we should comment why

cevian replied:
You are right. I cleaned up the logic here.

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,35 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+SELECT pg_catalog.pg_extension_config_dump(pg_get_serial_sequence('_timescaledb_catalog.bgw_job','id'), '');
+
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.bgw_job_stat (
+    job_id                  INT         PRIMARY KEY REFERENCES _timescaledb_catalog.bgw_job(id) ON UPDATE CASCADE ON DELETE CASCADE,
```
My reading of https://www.postgresql.org/docs/10/static/datatype-numeric.html#DATATYPE-SERIAL is that SERIAL == INTEGER, BIGSERIAL == BIGINT or am I missing something. Will change INT to INTEGER

cevian replied:
You are right removed the ON UPDATE CASCADE to be consistent with other catalog tables. The id should never be updated anyway.

cevian replied:
I think primary key is appropriate here for a 1:1 relation. A primary key is really a UNIQUE NOT NULL key. It's a matter of taste but I think PRIMARY KEY is better here.

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (!sjob->enabled || scheduled_bgw_job_is_running(sjob))
```
`scheduled_bgw_job_is_running` looks like it can cause deadlocks in the case where a background worker starts and immediately finishes before the scheduler gets a chance to check if the job is still alive. We should mention that in such a case the `WaitLatch` will be set by the default `SIGUSR1` handler.

cevian replied:
Not sure I understand scheduled_bgw_job_is_running checks a local variable. What do you mean?

JLockerman commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,345 @@
+}
+
+bool
+ts_bgw_job_stat_end_was_marked(BgwJobStat *jobstat)
```
This function name looks like it should be a getter for `ts_bgw_job_stat_mark_end`, maybe rename to something like `ts_bgw_job_stat_has_finish_timestamp`.

cevian replied:
I actually meant it as a getter for whether or not the "mark_end" operation was performed. Is there a better name you can suggest?

JLockerman replied:
Best I can think of is `.._was_finished` or `..._finished_cleanly` or something descriptive of how it should be used like that.

(The problem is that we have a lot of things starting with `ts_bgw_job_stat_...` split across both the struct and the tuple, in this case there's a `mark_end` in both, and one based on the other, so it's not obvious when each getter is valid. We may just have to accept confusing names. either way, we should document what the function means.)

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+
+static
+bool
+scheduled_bgw_job_list_start_workers()
```
I keep reading the function as scheduled_bgw_job_list_start**ed**_workers, but I don't know a better thing to name it. Adding a double underscore between the noun and verb, as in `scheduled_bgw_job_list__start_workers` would help, if we were willing to adopt that convention.

cevian replied:
I don't think we want to adopt that convention let's think about better names. Would `scheduled_bgw_job_list_launch_workers` be better?

JLockerman replied:
Yes, I think it's easier to understand.

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+static inline void
+bgw_scheduler_on_postmaster_death(void)
+{
+	/* don't call exit hooks cause we want to bail out quickly */
```
We need a comment on how: in the case of postmaster death, we no longer care about the state of shared memory, it'll be invalid.

cevian replied:
fixed

JLockerman commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+	TimestampTz next_start;
+	TimestampTz timeout_at;
+
+	bool		enabled;
```
Comment on which has higher priority, `enabled` or `next_start`.

cevian replied:
documented

JLockerman commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,345 @@
+
+	Interval	zero_ival = {.time = 0,};
+
+	values[Anum_bgw_job_stat_job_id - 1] = Int32GetDatum(bgw_job_id);
```
It would be nice to have a `AnumToIndex` macro so for self-documenting-code purposes.

JLockerman commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,345 @@
+
+static
+TimestampTz
+calculate_next_start_success(TimestampTz last_finish, BgwJob *job)
```
This function, and the next two, read like the they're calculating when the next successful/failing/crashing execution should run. Something like `calculate_next_start_on_success` might be clearer.

cevian replied:
That is better. changed.

JLockerman commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,345 @@
+	 * Unfortunaetly, 3 cannot be helped because when a scheduler gets a
+	 * SIGTERM it sends SIGTERMS to it's any running jobs as well. Since you
+	 * aren't supposed to write to the DB Once you get a sigterm, neither the
+	 * job nor the scheduler can mark the end of a job.
```
Could we handle this by adding a `terminated` field, which is unset by the scheduler? It seems unlikely that the background worker will be terminated directly, and in that case it probably is due to a crash.

cevian replied:
Not sure which case you are solving? If the background worker is terminated directly the scheduler already performs a mark_end. Can you explain. Upon a crash the scheduler is also sent a SIGQUIT so it cannot unset anything. Nor can it unset stuff if the scheduler is terminated.

JLockerman replied:
case 3, distinguishing between `SIGTERM` and background worker crashes by assuming the worker was terminated and having the scheduler mark the worker as not-terminated, if the scheduler is still alive.

JLockerman commented on src/bgw/README:
```diff
@@ -0,0 +1,34 @@
+TimescaleDB needas to run multiple background jobs. This module
+adds a simple scheduler so that jobs inserted into a jobs table
+could be run on a schedule. This first implementation has two limitations:
```
We need to describe the kinds of schedules we can run. At minimum we can run a job a fixed interval after the last time the job completed. Can we also run a job at a fixed rate?

cevian replied:
added a section

JLockerman commented on src/bgw/README:
```diff
@@ -0,0 +1,34 @@
+TimescaleDB needas to run multiple background jobs. This module
```
+1 to  READMEs and markdown.

Postgres has a convention of `README.foo` (e.g. [README.barrier](https://github.com/postgres/postgres/blob/520acab171244b55d816c70b9a89280b09937925/src/backend/storage/lmgr/README.barrier)) so they don't all have the same name, I think we should adopt something similar perhaps `README.foo.md`?

cevian replied:
Ok moving to markdown. @JLockerman I think we'll keep README.md at the top level of a module (in this case `bgw/README.md`) for the overall docs for the module. For any specialized docs we can use `README.foo.md`. Does that sound good?

JLockerman replied:
@cevian it’s reasonable, but it’ll make it a little more difficult to search for a specific file, and it’ll make tab names a bit harder to distinguish.

JLockerman replied:
(I think it’ll render better on github though, which probably makes it worthwhile)

JLockerman replied:
(I think it’ll render better on github though, which probably makes it worthwhile)

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,35 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.bgw_job (
+    id                  SERIAL      PRIMARY KEY,
+    job_name            NAME        NOT NULL,
+    job_type            NAME        NOT NULL,
```
I do not like the idea of using enum types since alter type add value cannot be executed inside a txn block and there is no alter type remove value at all (https://www.postgresql.org/docs/10/static/sql-altertype.html). This will make it hard to maintain in the future since we won't be able to use update scripts to change the set of values in an enum. We can consider creating a custom type where the in and out functions are defined in C and correspond to a C enum but I don't know if we want to go that far...

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,35 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+    job_name            NAME        NOT NULL,
+    job_type            NAME        NOT NULL,
+    schedule_interval   INTERVAL    NOT NULL,
+    job_timeout         INTERVAL    NOT NULL,
```
I used the job_prefix for things that are SQL keywords --- `name`, `type`, `timeout`. But changed `job_name`->`application_name` (which is the field in pg_stat_activity that this sets) and `job_timeout` to `max_runtime`. I kept `job_type` as is (couldn't think of a better name).

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,35 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
 
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.bgw_job (
+    id                  SERIAL      PRIMARY KEY,
+    job_name            NAME        NOT NULL,
```
done.

cevian commented on sql/tables.sql:
```diff
@@ -146,5 +146,35 @@ CREATE INDEX IF NOT EXISTS chunk_index_hypertable_id_hypertable_index_name_idx
+
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.bgw_job_stat (
+    job_id                  INT         PRIMARY KEY REFERENCES _timescaledb_catalog.bgw_job(id) ON UPDATE CASCADE ON DELETE CASCADE,
+    last_start              TIMESTAMPTZ NOT NULL DEFAULT NOW(),
```
Yeah I think this is a good signaling mechanism if nothing else.

cevian commented on src/bgw/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+set(SOURCES
+  "${CMAKE_CURRENT_SOURCE_DIR}/job.c"
```
Unfortunately it's not. The build breaks when I take it out. This is because of how `target_sources` works.

cevian commented on src/bgw/README:
```diff
@@ -0,0 +1,34 @@
+
+There is no prioritization for when to run jobs.
+
+There design of the scheduler is as follows:
```
fixed

cevian commented on src/extension.c:
```diff
@@ -248,3 +248,10 @@ extension_is_loaded(void)
+extension_get_so_name()
+{
+	/* TODO: after merge check whether this is the right place */
+	return EXTENSION_NAME "-" TIMESCALEDB_VERSION_MOD;
```
me neither. especially in a cross-platform way. leaving for now

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,348 @@
+	fd->total_runs++;
+
+	/*
+	 * This is udone by any of the end marks. This is so that we count crashes
```
fixed

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+	MemoryContext scheduler_mctx = CurrentMemoryContext;
+
+	StartTransactionCommand();
+	scheduled_jobs = scheduled_bgw_job_list_create(scheduler_mctx);
```
added comment

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+		long		timeout;
+		TimestampTz next_start;
+
+		if (quit_time < GetCurrentTimestamp())
```
fixed.

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+
+	Assert(sjob->enabled);
+
+	StartTransactionCommand();
```
Added function docs

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+				break;
+			case BGWH_STOPPED:
+				StartTransactionCommand();
+				scheduled_bgw_job_mark_stopped(sjob);
```
No because scheduled_bgw_job_mark_stopped will set the handle to NULL which will continue the loop above. I added a comment and Assert explaining this

JLockerman replied:
So compare this function to something like `waitpid`, that function will return information about the ground truth of the process state regardless of when and how it's called. This function returns information based on the last call to `scheduled_bgw_job_mark_stopped`, which, depending on where we are relative to `scheduled_bgw_job_list_check_stopped_or_timed_out_jobs` may or may not be up-to-date. A more accurate name would be something like `scheduled_bgw_job_is_scheduled`.

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,402 @@
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
```
This is actually related to the second limitation (prioritization) already in the README.

erimatnor commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+				 tuple_found_func tuple_found, tuple_filter_func tuple_filter, void *data, MemoryContext mctx, LOCKMODE lockmode)
+{
+	Catalog    *catalog = catalog_get();
+	ScannerCtx	scanCtx = {
```
scanctx

cevian replied:
fixed

erimatnor commented on src/bgw/job_stat.h:
```diff
@@ -0,0 +1,29 @@
+
+typedef struct BgwJobStat
+{
+	FormData_bgw_job_stat fd;
```
Add `JobType` to this struct and set when you read the tuple.

Then we need not call `get_job_type_from_name()` all the time.

cevian replied:
fixed

erimatnor commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+bool
+bgw_job_execute(BgwJob *job)
+{
+	switch (get_job_type_from_name(&job->fd.job_type))
```
A switch is fine, but if the function interface is the same for all jobs, you can just create a static type-to-function array and index that as an alternative approach.

Also, see my note at the definition of `BgwJob` on caching the type in the struct so that you can do `job->type`.

cevian replied:
Added caching. I like the switch more for now.

erimatnor commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,386 @@
+
+
+	sjob->handle = bgw_job_start(&sjob->job);
+	sjob->state = JOB_STATE_STARTED;
```
Shouldn't there be a single function to transition to `JOB_STATE_STARTED`, which also sets the state? I.e., this state should be set in `bgw_job_start()`, no?

cevian replied:
Fixed

erimatnor commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,386 @@
+	JOB_STATE_DISABLED,
+
+	/* next states: JOB_STATE_STARTED */
+	JOB_STATE_WAITING,
```
Call this NOT_RUNNING or something since WAITING sounds like the job is running and waiting on something.

I think we decided on:

DISABLED
SCHEDULED
STARTED
TERMINATING -> SCHEDULED | DISABLED

Add some comments on what each state means, i.e., what the scheduler vs. job believes.

cevian replied:
Fixed

erimatnor commented on test/expected/insert.out:
```diff
@@ -38,89 +38,103 @@ INSERT INTO "two_Partitions"("timeCustom", device_id, series_0, series_1) VALUES
- _timescaledb_internal."_hyper_1_4_chunk_two_Partitions_timeCustom_series_bool_idx" | i    | timeCustom  | bigint           | f
- _timescaledb_internal."_hyper_1_4_chunk_two_Partitions_timeCustom_series_bool_idx" | i    | series_bool | boolean          | f
-(76 rows)
+                                      Relation                                      | Kind |        Column        |       Column type        | Nullable 
```
Should probably do ` test.show_columnsp('_timescaledb_internal._hyper_%');` here to avoid outputting job_stats tables.

cevian replied:
Fixed

erimatnor commented on test/expected/drop_rename_hypertable.out:
```diff
@@ -30,85 +30,98 @@ INSERT INTO "two_Partitions"("timeCustom", device_id, series_0, series_1) VALUES
- _timescaledb_internal."_hyper_1_4_chunk_two_Partitions_timeCustom_series_bool_idx" | i    | timeCustom  | bigint           | f
- _timescaledb_internal."_hyper_1_4_chunk_two_Partitions_timeCustom_series_bool_idx" | i    | series_bool | boolean          | f
-(76 rows)
+                                      Relation                                      | Kind |        Column        |       Column type        | Nullable 
```
should probably do `test.show_columnsp('_timescaledb_internal._hyper_%');` here to avoid outputting job_stats tables

cevian replied:
Fixed

erimatnor commented on test/expected/drop_chunks.out:
```diff
@@ -54,10 +54,11 @@ WHERE h.schema_name = 'public' AND (h.table_name = 'drop_chunk_test1' OR h.table
- Schema | Name | Type | Owner 
---------+------+------+-------
-(0 rows)
+                     List of relations
```
should we filter the`bgw_job_stat` output in this test? It is not related to the test as we only want to see chunks.

cevian replied:
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+};
+
+
+#define VERSION_CHECK_JOB_NAME "check_for_update_if_enabled"
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+get_job_type_from_name(Name job_type_name)
+{
+	for (int i = 0; i < _MAX_JOB_TYPE; i++)
+	{
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+}
+
+static
+JobType
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+
+#ifdef DEBUG
+#define TEST_JOB_FUNCTION "ts_bgw_job_execute_test"
+static bool
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+			/* TODO: fill in when merging with Amy's stuff  */
+			return false;
+		case _MAX_JOB_TYPE:
+			/* unkown_job */
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+#else
+			elog(ERROR, "unknown job type \"%s\"", NameStr(job->fd.job_type));
+#endif
+			break;
```
fixed

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,317 @@
+
+#define VERSION_CHECK_JOB_NAME "check_for_update_if_enabled"
+
+TS_FUNCTION_INFO_V1(ts_bgw_job_entrypoint);
```
fixed

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,359 @@
+*  a minimum wait after a crash that we wait, so that if an operator needs to disable the job,
+*  there will be enough time before another crash.
+*/
+static
```
fixed

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,359 @@
+		return min_time;
+	return failure_calc;
+}
+static bool
```
fixed

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,386 @@
+}
+
+static
+List *
```
fixed

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,386 @@
+}
+
+/*
+*	This function starts a job.
```
fixed

cevian commented on src/bgw/scheduler.c:
```diff
@@ -0,0 +1,386 @@
+	check_for_stopped_and_timed_out_jobs();
+}
+
+static
```
Fixed

cevian commented on src/bgw/timer.c:
```diff
@@ -0,0 +1,110 @@
+
+
+#define MAX_TIMEOUT (5*1000L)
+static inline void
```
fixed

cevian commented on src/bgw/timer.c:
```diff
@@ -0,0 +1,110 @@
+}
+
+
+static const Timer StandardTimer = {
```
fixed

cevian commented on test/src/bgw/params.c:
```diff
@@ -0,0 +1,247 @@
+	return get_relname_relid("bgw_dsm_handle_store", get_namespace_oid("public", false));
+}
+
+static
```
fixed

cevian commented on test/src/bgw/params.c:
```diff
@@ -0,0 +1,247 @@
+	slock_t		mutex;
+} TestParamsWrapper;
+
+static
```
fixed

cevian commented on test/src/bgw/params.c:
```diff
@@ -0,0 +1,247 @@
+	return handle;
+}
+
+static
```
fixed

cevian commented on test/src/bgw/params.c:
```diff
@@ -0,0 +1,247 @@
+	return wrapper;
+};
+
+static
```
fixed

cevian commented on test/src/bgw/params.c:
```diff
@@ -0,0 +1,247 @@
+	params_close_wrapper(wrapper);
+}
+
+static
```
fixed

cevian commented on test/src/bgw/scheduler_mock.c:
```diff
@@ -0,0 +1,268 @@
+	[TEST_JOB_TYPE_JOB_3_LONG] = "bgw_test_job_3_long",
+};
+
+static
```
fixed

cevian commented on test/src/bgw/scheduler_mock.c:
```diff
@@ -0,0 +1,268 @@
+}
+
+static
+bool
```
fixed

cevian commented on test/src/bgw/scheduler_mock.c:
```diff
@@ -0,0 +1,268 @@
+}
+
+static
+bool
```
fixed

cevian commented on test/src/bgw/scheduler_mock.c:
```diff
@@ -0,0 +1,268 @@
+	return true;
+}
+
+static
```
fixed

cevian commented on test/src/bgw/timer_mock.h:
```diff
@@ -0,0 +1,22 @@
+	.get_current_timestamp = mock_current_time,
+	.wait = mock_wait,
+};
+
```
fixed

cevian commented on test/src/bgw/timer_mock.h:
```diff
@@ -0,0 +1,22 @@
+extern void timer_mock_register_bgw_handle(BackgroundWorkerHandle *handle);
+extern TimestampTz mock_current_time(void);
+
+static const Timer MockTimer = {
```
Fixed made it a const and defined in timer_mock.c

cevian commented on test/src/bgw/scheduler_mock.c:
```diff
@@ -0,0 +1,268 @@
+
+	worker_handle = bgw_start_worker("ts_bgw_db_scheduler_test_main", "ts_bgw_db_scheduler_test_main", params);
+
+	WaitForBackgroundWorkerStartup(worker_handle, &pid);
```
Added Asserts here. I think asserts are better in tests as they make it easier to debug (failure produces a core dump).

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+(1 row)
+
+\c single :ROLE_DEFAULT_PERM_USER
+--should run it the first time
```
Fixed

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+ Running mock wait until 25000, started at 0
+(2 rows)
+
+--error job should run 1 more times and increases the wait
```
Fixed

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+SELECT * FROM _timescaledb_internal.bgw_job_stat;
+ job_id | last_start | last_finish | next_start | last_run_success | total_runs | total_duration | total_successes | total_failures | total_crashes | consecutive_failures | consecutive_crashes 
+--------+------------+-------------+------------+------------------+------------+----------------+-----------------+----------------+---------------+----------------------+---------------------
+(0 rows)
```
fixed

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+ 
+(1 row)
+
+SELECT ts_bgw_db_scheduler_test_run_and_wait(50);
```
Fixed naming of function

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+                     msg                     
+---------------------------------------------
+ registering new background worker
+ Running mock wait until 25000, started at 0
```
Fixed

cevian commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,603 @@
+SELECT * FROM bgw_log;
+                     msg                     
+---------------------------------------------
+ registering new background worker
```
Changed a bunch of this. Made it hook into the elog system which works much better. Added mock time and application name fields.

davidkohn88 commented on src/bgw/README.md:
```diff
@@ -0,0 +1,47 @@
+
+TimescaleDB needs to run multiple background jobs. This module
+adds a simple scheduler so that jobs inserted into a jobs table
+could be run on a schedule.
```
could-> can

JLockerman commented on test/sql/bgw_scheduler.sql:
```diff
@@ -0,0 +1,295 @@
+    r INTEGER;
+BEGIN
+
+    FOR i in 1..10
```
Doc why only attempt this 10 times.

davidkohn88 commented on src/bgw/job.c:
```diff
@@ -0,0 +1,250 @@
+
+	elog(DEBUG1, "started background job %d", job_id);
+
+	BackgroundWorkerInitializeConnectionByOid(db_oid, InvalidOid);
```
Given that we've already connected to the DB as the scheduler, it seems highly unlikely it will have disappeared before we run this bit. In fact, I think it might be impossible, given that if you drop the db, then we'll stop the worker etc. We haven't yet addressed the potential problems with eventually wanting to be able to run jobs as not a superuser, but that's a different story for a different day.

davidkohn88 commented on src/bgw/job.c:
```diff
@@ -0,0 +1,311 @@
+	BackgroundWorkerInitializeConnectionByOid(db_oid, InvalidOid);
+
+	/*
+	 * Start transactions early. Jobs are allowed to end and start transaction
```
I think this comment needs to be updated given that we aren't doing the thing where the txn stays open for the duration of the job anymore.

cevian replied:
fixed

JLockerman commented on test/src/bgw/timer_mock.c:
```diff
@@ -0,0 +1,57 @@
+		WaitForBackgroundWorkerShutdown(bgw_handle);
+		bgw_handle = NULL;
+	}
+	params_set_time(until);
```
I think we need a `params_set_time_on_less_than` to allow background workers to wait for a timeout, otherwise we can get the timetravel if a process waiting for a lower `until` races one with a greater `until`. Until we implement something like that we need a warning on this function, and `mock_timer` in the header like

We need a comment on this function like

```C
/* WARNING: mock_wait must _only_ be called from the bgw_scheduler, calling it from a worker will clobber the timer state */
```

davidkohn88 commented on test/expected/bgw_scheduler.out:
```diff
@@ -0,0 +1,744 @@
+(1 row)
+
+INSERT INTO _timescaledb_config.bgw_job (application_name, job_type, schedule_interval, max_runtime, max_retries, retry_period) VALUES
+('test_job_1', 'bgw_test_job_1', Interval '100ms', Interval '100s', 3, Interval '1s');
```
Interval -> interval or INTERVAL 
title case is just odd.

cevian commented on src/bgw/job.h:
```diff
@@ -0,0 +1,33 @@
+typedef struct BgwJob
+{
+	FormData_bgw_job fd;
+	JobType		bgw_type;
```
pg_indent won't like that

cevian commented on src/bgw/job_stat.c:
```diff
@@ -0,0 +1,359 @@
+ * along with a ceiling at schedule_interval * MAX_INTERVALS_BACKOFF */
+static
+TimestampTz
+calculate_next_start_on_failure(TimestampTz last_finish, int consecutive_failures, BgwJob *job)
```
I explicitly wanted these functions with simple parameters so that they would be easier to test.

cevian commented on src/bgw/job.c:
```diff
@@ -0,0 +1,311 @@
+}
+
+
+TS_FUNCTION_INFO_V1(ts_bgw_job_entrypoint);
```
This is now an actual entrypoint used for the bgw launcher so leaving

davidkohn88 replied:
Still it doesn't need to be exported that way.
